bugsnag-android-core/src/main/java/com/bugsnag/android/internal/Journal.kt

```diff
@@ -0,0 +1,155 @@
+package com.bugsnag.android.internal
+
+import com.bugsnag.android.Logger
+import com.dslplatform.json.DslJson
+import java.io.IOException
+import java.io.OutputStream
+
+/**
+ * A journal represents a series of commands that will manipulate a document.
+ * Sometimes it's more efficient to serialize a base document and a list of modifications
+ * to be made to it rather than to re-serialize the entire document after every modification.
+ */
+class Journal {
+    private val commands: MutableList<Command> = mutableListOf()
+
+    /**
+     * Apply this journal to a document.
+     *
+     * This method goes through the journal entries one-by-one, applying their modification
+     * commands to the document. This is a destructive operation that will modify the document.
+     *
+     * Note: If a journal command replaces the root object, the resulting document will be a
+     * different object from the one passed in.
+     *
+     * @param document The document to apply to.
+     * @return The finished document (may be a different object from the one passed in).
+     */
+    fun applyTo(document: MutableMap<in String, out Any>): MutableMap<in String, out Any> {
+        var updatedDocument = document
+        for (command in commands) {
+            updatedDocument = command.apply(updatedDocument)
+        }
+        return updatedDocument
+    }
+
+    /**
+     * Add a journal command to this journal
+     *
+     * @param command The journal command
+     */
+    fun add(command: Command) {
+        commands.add(command)
+    }
+
+    /**
+     * Clear the journal, removing all journal entries.
+     */
+    fun clear() {
+        commands.clear()
+    }
+
+    /**
+     * Serialize this journal to a stream.
+     *
+     * @param out The stream to serialize to.
+     * @throws IOException if an IO exception occurs.
+     */
+    @Throws(IOException::class)
+    fun serialize(out: OutputStream) {
+        for (command in commands) {
+            command.serialize(out)
+        }
+    }
+
+    /**
+     * Command represents a single journal command, consisting of:
+     * - a document path, which determines where in a document the change is to be made.
+     * - a value, which determines the change to be made.
+     *
+     * In general, a null value represents a delete, and a non-null value represents an insert or
+     * replace.
+     *
+     * @see DocumentPath for a description of how path strings work.
+     */
+    class Command(private val path: String, val value: Any?) {
+        private val documentPath: DocumentPath = DocumentPath(path)
```

Comment:
`DocumentPath` only uses the string representation to initialize itself. After that it has no use for it. However, `JournalEntry` needs it for serialization purposes. To me it makes more sense to store it in `JournalEntry` rather than putting it in `DocumentPath` only to reach into the object later.